File: mobile/shared/src/commonMain/kotlin/com/bissbilanz/repository/WeightRepository.kt
package com.bissbilanz.repository

import app.cash.sqldelight.coroutines.asFlow
import app.cash.sqldelight.coroutines.mapToList
import com.bissbilanz.ErrorReporter
import com.bissbilanz.analytics.WeightChartInput
import com.bissbilanz.analytics.weightMovingAverage
import com.bissbilanz.api.BissbilanzApi
import com.bissbilanz.api.generated.model.WeightCreate
import com.bissbilanz.api.generated.model.WeightEntry
import com.bissbilanz.api.generated.model.WeightTrendEntry
import com.bissbilanz.api.generated.model.WeightUpdate
import com.bissbilanz.cache.BissbilanzDatabase
import com.bissbilanz.mode.AppModeManager
import com.bissbilanz.sync.SyncOperation
import com.bissbilanz.sync.SyncQueue
import com.bissbilanz.sync.rewriteQueuedCreate
import com.bissbilanz.userdata.UserDataDatabase
import com.bissbilanz.util.decodeOrNull
import com.bissbilanz.util.isTempId
import com.bissbilanz.util.newTempId
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.IO
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlin.time.Clock

class WeightRepository(
    private val api: BissbilanzApi,
    private val db: UserDataDatabase,
    private val cacheDb: BissbilanzDatabase,
    private val syncQueue: SyncQueue,
    private val json: Json,
    private val errorReporter: ErrorReporter,
    private val appModeManager: AppModeManager,
) {
    var onWeightChanged: (suspend () -> Unit)? = null

    /**
     * Fired after a server refresh replaced the cached entries. Weights logged
     * outside this device (web, MCP, iOS) only arrive through refresh, so this is
     * where the Health Connect export, the weight widget and the watch state catch
     * them — [onWeightChanged] only covers mutations made in the app itself.
     * Mirrors [com.bissbilanz.repository.EntryRepository.onEntriesRefreshed].
     */
    var onWeightRefreshed: (suspend () -> Unit)? = null

    fun entries(): Flow<List<WeightEntry>> =
        db.userDataDatabaseQueries
            .selectAllWeightEntries()
            .asFlow()
            .mapToList(Dispatchers.IO)
            .map { rows -> rows.mapNotNull { json.decodeOrNull<WeightEntry>(it.jsonData) } }

    suspend fun refresh(limit: Int = 30) {
        if (appModeManager.isLocal) return
        val entries = api.getWeightEntries(limit)
        cacheWeightEntries(entries)
        onWeightRefreshed?.invoke()
    }

    suspend fun createEntry(entry: WeightCreate): WeightEntry {
        val temp = weightCreateToEntry(entry)
        cacheWeightEntry(temp)
        syncQueue.enqueue(SyncOperation.CreateWeight(json.encodeToString(entry), localId = temp.id))
        onWeightChanged?.invoke()
        return temp
    }

    suspend fun updateEntry(
        id: String,
        entry: WeightUpdate,
    ): WeightEntry {
        val cached = db.userDataDatabaseQueries.selectAllWeightEntries().executeAsList()
        val existing = cached.mapNotNull { json.decodeOrNull<WeightEntry>(it.jsonData) }.find { it.id == id }
        val result =
            if (existing != null) {
                val updated =
                    existing.copy(
                        weightKg = entry.weightKg ?: existing.weightKg,
                        entryDate = entry.entryDate ?: existing.entryDate,
                        notes = entry.notes ?: existing.notes,
                    )
                cacheWeightEntry(updated)
                updated
            } else {
                WeightEntry(
                    id = id,
                    userId = "",
                    weightKg = entry.weightKg ?: 0.0,
                    entryDate = entry.entryDate ?: "",
                    notes = entry.notes,
                )
            }
        if (id.isTempId()) {
            coalesceQueuedCreate(id, entry)
        } else {
            syncQueue.enqueue(SyncOperation.UpdateWeight(id, json.encodeToString(entry)))
        }
        onWeightChanged?.invoke()
        return result
    }

    /**
     * Rewrites the still-queued Create operation for a temp-id weight entry so the
     * eventual upload carries the edited values. If the create has already been drained
     * (no queued op found), the update is skipped — the temp id is unknown server-side.
     */
    private suspend fun coalesceQueuedCreate(
        tempId: String,
        update: WeightUpdate,
    ) {
        syncQueue.rewriteQueuedCreate("weight", tempId) { op ->
            val create = op as? SyncOperation.CreateWeight ?: return@rewriteQueuedCreate null
            val body = json.decodeOrNull<WeightCreate>(create.body) ?: return@rewriteQueuedCreate null
            val merged =
                body.copy(
                    weightKg = update.weightKg ?: body.weightKg,
                    entryDate = update.entryDate ?: body.entryDate,
                    notes = update.notes ?: body.notes,
                )
            create.copy(body = json.encodeToString(merged))
        }
    }

    suspend fun getTrend(
        from: String,
        to: String,
    ): List<WeightTrendEntry> {
        if (appModeManager.isLocal) return trendFromCache(from, to)
        return try {
            api.getWeightTrend(from, to)
        } catch (e: Exception) {
            if (e is kotlinx.coroutines.CancellationException) throw e
            errorReporter.captureException(e)
            trendFromCache(from, to)
        }
    }

    private fun trendFromCache(
        from: String,
        to: String,
    ): List<WeightTrendEntry> =
        db.userDataDatabaseQueries
            .selectAllWeightEntries()
            .executeAsList()
            .mapNotNull { json.decodeOrNull<WeightEntry>(it.jsonData) }
            .filter { it.entryDate in from..to }
            .map { WeightChartInput(date = it.entryDate, weightKg = it.weightKg, loggedAt = it.loggedAt) }
            .let(::weightMovingAverage)
            .map { WeightTrendEntry(entryDate = it.date, weightKg = it.weightKg, movingAvg = it.movingAvg) }

    suspend fun deleteEntry(id: String) {
        db.userDataDatabaseQueries.deleteWeightEntry(id)
        if (id.isTempId()) {
            syncQueue.removeByAffected("weight", id)
        } else {
            syncQueue.enqueue(SyncOperation.DeleteWeight(id))
        }
        onWeightChanged?.invoke()
    }

    private fun cacheWeightEntry(entry: WeightEntry) {
        db.userDataDatabaseQueries.insertWeightEntry(
            id = entry.id,
            entryDate = entry.entryDate,
            weightKg = entry.weightKg,
            loggedAt = entry.loggedAt,
            jsonData = json.encodeToString(entry),
        )
    }

    private suspend fun cacheWeightEntries(entries: List<WeightEntry>) {
        val pendingIds = pendingWeightIds()
        val queries = db.userDataDatabaseQueries
        // Keep optimistic temp-id creates whose upload is still queued or in flight,
        // and rows carrying a queued update (a queued delete already removed its row).
        // A forced refresh right after a weight create/edit/delete (the UI calls
        // refresh() on save) races the async sync-queue upload; without this, the
        // still-stale server list would revert the edit, drop the just-logged entry,
        // or resurrect a deleted one until the next manual refresh. A temp row that is
        // no longer in `pendingIds` has nothing left to upload — the sync manager
        // already swapped it for the server record — so it must not be kept, or the
        // list shows the same weight twice.
        val preserved =
            queries
                .selectAllWeightEntries()
                .executeAsList()
                .filter { it.id in pendingIds }
                .mapNotNull { json.decodeOrNull<WeightEntry>(it.jsonData) }
        queries.transaction {
            queries.deleteAllWeightEntries()
            entries.forEach { entry -> if (entry.id !in pendingIds) cacheWeightEntry(entry) }
            preserved.forEach { cacheWeightEntry(it) }
        }
        // SyncMeta lives in the cache database; written after the user-data commit.
        cacheDb.bissbilanzDatabaseQueries.upsertSyncMeta(
            entityType = "weight",
            lastSyncedAt = Clock.System.now().toString(),
        )
    }

    /** Weight-entry ids with an un-uploaded (queued or in-flight) sync operation. */
    private suspend fun pendingWeightIds(): Set<String> =
        syncQueue
            .all()
            .asSequence()
            .filter { it.operation.affectedTable == "weight" }
            .mapNotNull { it.operation.affectedId }
            .toSet()

    private fun weightCreateToEntry(entry: WeightCreate): WeightEntry =
        WeightEntry(
            id = newTempId(),
            userId = "",
            weightKg = entry.weightKg,
            entryDate = entry.entryDate,
            loggedAt = Clock.System.now().toString(),
            notes = entry.notes,
            createdAt = Clock.System.now().toString(),
        )
}

File: mobile/iosApp/Bissbilanz/Persistence/WeightRepository.swift
import Foundation
import Observation
import SwiftData

/// Local-first repository for weight entries. Reads come from SwiftData
/// (sorted by entry date, newest first); `refresh()` upserts the server list
/// by id and drops rows deleted elsewhere. Writes are SwiftData-first with
/// the upload queued via the sync manager. Server-computed weight stats stay
/// on the direct API in the views (hidden in Local mode).
@MainActor
@Observable
final class WeightRepository {
    private let context: ModelContext
    private let api: BissbilanzAPI
    private let appMode: AppModeManager
    private let syncManager: SyncManager

    init(context: ModelContext, api: BissbilanzAPI, appMode: AppModeManager, syncManager: SyncManager) {
        self.context = context
        self.api = api
        self.appMode = appMode
        self.syncManager = syncManager
    }

    // MARK: - Reads (local)

    func entries() -> [WeightEntry] {
        let descriptor = FetchDescriptor<LocalWeightEntry>(sortBy: [
            SortDescriptor(\.entryDate, order: .reverse),
        ])
        let rows = (try? context.fetch(descriptor)) ?? []
        return rows.compactMap { $0.toWeightEntry() }
    }

    /// One page of entries, newest first — backs the paginated history list.
    func entries(offset: Int, limit: Int) -> [WeightEntry] {
        var descriptor = FetchDescriptor<LocalWeightEntry>(sortBy: [
            SortDescriptor(\.entryDate, order: .reverse),
        ])
        descriptor.fetchOffset = offset
        descriptor.fetchLimit = limit
        let rows = (try? context.fetch(descriptor)) ?? []
        return rows.compactMap { $0.toWeightEntry() }
    }

    /// Just the logged days, read off the typed column. The Health import only
    /// needs to know which days are already taken, and `entries()` would JSON-
    /// decode every stored row to answer that — on every foreground activation.
    func entryDates() -> Set<String> {
        let rows = (try? context.fetch(FetchDescriptor<LocalWeightEntry>())) ?? []
        return Set(rows.map(\.entryDate))
    }

    func latest() -> WeightEntry? {
        var descriptor = FetchDescriptor<LocalWeightEntry>(sortBy: [
            SortDescriptor(\.entryDate, order: .reverse),
        ])
        descriptor.fetchLimit = 1
        return (try? context.fetch(descriptor))?.first?.toWeightEntry()
    }

    /// The entry whose day is nearest to `date` ("yyyy-MM-dd") in either
    /// direction, preferring the on-or-before side on a tie.
    func closest(to date: String) -> WeightEntry? {
        var beforeDescriptor = FetchDescriptor<LocalWeightEntry>(
            predicate: #Predicate { $0.entryDate <= date },
            sortBy: [SortDescriptor(\.entryDate, order: .reverse)]
        )
        beforeDescriptor.fetchLimit = 1
        var afterDescriptor = FetchDescriptor<LocalWeightEntry>(
            predicate: #Predicate { $0.entryDate > date },
            sortBy: [SortDescriptor(\.entryDate, order: .forward)]
        )
        afterDescriptor.fetchLimit = 1
        let before = (try? context.fetch(beforeDescriptor))?.first?.toWeightEntry()
        let after = (try? context.fetch(afterDescriptor))?.first?.toWeightEntry()
        guard let before else { return after }
        guard let after else { return before }
        return DateFormatting.dayDistance(before.entryDate, date)
            <= DateFormatting.dayDistance(after.entryDate, date) ? before : after
    }

    // MARK: - Refresh (API → store)

    /// One bulk fetch of the stored rows, keyed by id, instead of decoding every
    /// row through `entries()` just to read its id and then running a
    /// `fetchRow(id:)` per server row on top. That was an N+1 fetch plus a
    /// decode and a re-encode for every historical entry, on the main actor,
    /// several times a minute while browsing days — `DashboardView.loadData`
    /// calls this on every day swipe, sheet dismissal and pull-to-refresh, and
    /// `BackgroundRefresher.pull` calls it again on every background run.
    func refresh() async throws {
        guard !appMode.isLocal else { return }
        let fetched = try await api.getWeightEntries()
        let serverIds = Set(fetched.map(\.id))
        // Rows with an un-uploaded queued write must survive the server
        // response: a refresh racing the sync-queue upload would otherwise
        // reapply the stale server copy over the user's edit (see
        // EntryRepository.refresh, PR #416).
        let pendingIds = syncManager.pendingAffectedIds(table: "weight")
        // Pull `jsonData` eagerly: the per-row change check below reads it, and
        // on a faulted row that read is a separate SQLite trip per entry.
        var rowsDescriptor = FetchDescriptor<LocalWeightEntry>()
        rowsDescriptor.propertiesToFetch = [\.id, \.jsonData]
        var rowsById = Dictionary(
            ((try? context.fetch(rowsDescriptor)) ?? []).map { ($0.id, $0) },
            uniquingKeysWith: { first, _ in first }
        )
        for (id, row) in rowsById where !serverIds.contains(id)
            && !LocalStore.isTempId(id) && !pendingIds.contains(id)
        {
            context.delete(row)
            rowsById.removeValue(forKey: id)
        }
        for entry in fetched where !pendingIds.contains(entry.id) {
            if let row = rowsById[entry.id] {
                // Only dirty rows that actually changed. Rewriting every row
                // made each `save()` flush the whole table on the main thread,
                // which is what the SleepRepository.save hang in Sentry
                // BISSBILANZ-34 was.
                if row.jsonData != LocalStoreCoding.encode(entry) {
                    row.update(from: entry)
                }
            } else {
                let row = LocalWeightEntry(entry: entry)
                context.insert(row)
                rowsById[entry.id] = row
            }
        }
        save()
    }

    // MARK: - Writes (local first + queued upload)

    @discardableResult
    func createEntry(_ create: WeightCreate) async throws -> WeightEntry {
        let temp = makeEntry(from: create, id: LocalStore.makeTempId())
        upsert(temp)
        save()
        syncManager.enqueue(.createWeight(body: create, localId: temp.id))
        return temp
    }

    /// A missing local row means the record is gone (deleted elsewhere, or
    /// never cached) — the failure is reported without also queueing an upload
    /// for it. Enqueuing first and throwing afterwards told the caller the edit
    /// had failed while it was already on its way to the server, so the next
    /// refresh brought back the change the UI had just reported as failed.
    @discardableResult
    func updateEntry(id: String, _ update: WeightUpdate) async throws -> WeightEntry {
        guard let row = fetchRow(id: id), let existing = row.toWeightEntry() else {
            throw APIError.notFound
        }
        let patch = (try? JSONPatch.dictionary(of: update)) ?? [:]
        let updated = (try? JSONPatch.merged(WeightEntry.self, base: existing, patch: patch)) ?? existing
        row.update(from: updated)
        save()
        if LocalStore.isTempId(id) {
            coalesceQueuedCreate(tempId: id, update: update)
        } else {
            syncManager.enqueue(.updateWeight(id: id, body: update))
        }
        return updated
    }

    func deleteEntry(id: String) async throws {
        deleteRow(id: id)
        save()
        if LocalStore.isTempId(id) {
            syncManager.removeQueued(table: "weight", affectedId: id)
        } else {
            syncManager.enqueue(.deleteWeight(id: id))
        }
    }

    /// Rewrites the still-queued create for a temp-id weight entry so the
    /// eventual upload carries the edited values.
    private func coalesceQueuedCreate(tempId: String, update: WeightUpdate) {
        for row in syncManager.queuedOperations(table: "weight", affectedId: tempId) {
            guard let operation = row.operation(),
                  case let .createWeight(body, localId) = operation
            else { continue }
            let patch = (try? JSONPatch.dictionary(of: update)) ?? [:]
            let merged = (try? JSONPatch.merged(WeightCreate.self, base: body, patch: patch)) ?? body
            syncManager.replace(row, with: .createWeight(body: merged, localId: localId))
        }
    }

    // MARK: - Conversion helpers

    private func makeEntry(from create: WeightCreate, id: String) -> WeightEntry {
        let now = DateFormatting.isoDateTimeString(from: Date())
        return WeightEntry(
            id: id,
            userId: "",
            weightKg: create.weightKg,
            entryDate: create.entryDate,
            loggedAt: now,
            notes: create.notes,
            createdAt: now,
            updatedAt: nil
        )
    }

    // MARK: - Store helpers

    private func fetchRow(id: String) -> LocalWeightEntry? {
        var descriptor = FetchDescriptor<LocalWeightEntry>(predicate: #Predicate { $0.id == id })
        descriptor.fetchLimit = 1
        return (try? context.fetch(descriptor))?.first
    }

    private func upsert(_ entry: WeightEntry) {
        if let row = fetchRow(id: entry.id) {
            row.update(from: entry)
        } else {
            context.insert(LocalWeightEntry(entry: entry))
        }
    }

    private func deleteRow(id: String) {
        if let row = fetchRow(id: id) {
            context.delete(row)
        }
    }

    private func save() {
        try? context.save()
        WidgetSnapshotWriter.scheduleUpdate(context: context)
        syncLatestToHealth()
    }

    // MARK: - Apple Health write-back

    /// Dates `HealthKitImporter` is currently writing entries for, read *from*
    /// Health, with the number of import scopes currently covering each. Scoped
    /// to dates rather than a global flag so a refresh landing a weight for some
    /// other day mid-import still writes back normally.
    @ObservationIgnored private var healthImportDateCounts: [String: Int] = [:]

    /// Runs `body` with the Health write-back suppressed for `dates`. Those
    /// entries came out of Health, so they get marked as already synced instead
    /// of being written back as app-authored duplicates of the scale's samples.
    ///
    /// Reference-counted, because overlapping imports are the normal case:
    /// `BissbilanzApp` runs `importAllIfEnabled` on every foreground
    /// activation while the Weight tab independently imports from its own `.task`
    /// and pull-to-refresh. With a plain set cleared in `defer`, the inner
    /// scope's exit re-enabled write-back while the outer one was still
    /// writing — producing exactly the app-authored duplicates of the
    /// device's own samples that this suppression exists to prevent.
    func withHealthImportInProgress<T>(
        dates: Set<String>,
        _ body: @MainActor () async throws -> T
    ) async rethrows -> T {
        for date in dates {
            healthImportDateCounts[date, default: 0] += 1
        }
        defer {
            for date in dates {
                guard let count = healthImportDateCounts[date] else { continue }
                if count <= 1 {
                    healthImportDateCounts.removeValue(forKey: date)
                } else {
                    healthImportDateCounts[date] = count - 1
                }
            }
        }
        return try await body()
    }

    /// Pushes the newest weight to Apple Health on every store mutation — writes,
    /// refreshes and deletes alike. This used to live in the weight edit sheet,
    /// which meant a weight logged on the Apple Watch, or arriving from web, MCP
    /// or Android through `refresh()`, never reached Health at all. Android hangs
    /// the same export off its repository callbacks.
    ///
    /// Fire-and-forget, like the widget update above it. The marker guard inside
    /// `syncLatestWeight` makes an unchanged repeat free.
    private func syncLatestToHealth() {
        guard let latest = latest() else { return }
        let alreadyInHealth = healthImportDateCounts[latest.entryDate] != nil
        Task {
            await HealthKitService.shared.syncLatestWeight(
                latest.weightKg,
                entryDate: latest.entryDate,
                alreadyInHealth: alreadyInHealth
            )
        }
    }
}
